apps/server-asset-sg/src/utils/process-queue.ts

```diff
@@ -0,0 +1,82 @@
+import { sleep } from '@asset-sg/shared/v2';
+
+/**
+ * `ProcessQueue` is an async task runner that limits the number of tasks that can be run in parallel.
+ */
+export class ProcessQueue {
+  private readonly buffer: Process[] = [];
+  private readonly active = new Map<Process, Promise<void>>();
+
+  constructor(private readonly quantity: number) {
+    if (this.quantity < 1) {
+      throw new Error('Quantity must a positive integer.');
+    }
+  }
+
+  /**
+   * Registers a new tasks to be run.
+   *
+   * @param run The task to run.
+   */
+  add(run: () => void | Promise<void>): Promise<void> {
+    if (this.active.size < this.quantity) {
+      return new Promise((resolve) => {
+        const process: Process = { run, resolve };
+        this.run(process).then();
+      });
+    }
+    return new Promise((resolve) => {
+      const process: Process = { run, resolve };
+      this.buffer.push(process);
+    });
+  }
+
+  /**
+   * Returns a promise that resolves when the queue has no more tasks to run.
+   */
+  async waitForIdle(): Promise<void> {
+    while (true) {
+      if (this.active.size === 0 && this.buffer.length === 0) {
+        return;
+      }
+      if (this.active.size > 0) {
+        await Promise.all(this.active.values());
+      } else {
+        await sleep(1);
+      }
+    }
+  }
+
+  /**
+   * Runs a task. Automatically continues to run any queued tasks afterward.
+   *
+   * @param process The task to run.
+   * @private
+   */
+  private async run(process: Process): Promise<void> {
+    const promise = (process.run() as Promise<void> | undefined) ?? Promise.resolve();
```

Comment:
Maybe use instanceof to not have to rely on type assertions and improve readability (for the sake of simplicity):

```
const run = process.run()
const promise = run instanceof Promise ? run : Promise.resolve();
```

Not sure about timing issues, but I don't think they'd matter in this context.